File: languages/cpp/02-core/sorting-and-searching/example/stable-vs-unstable-sort.cpp
// This helper example focuses on contrasting two sorting behaviors without extra setup noise.

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

struct Record {
    int score;
    string name;
};

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
void printRecords(const vector<Record>& records, const string& title) {
    cout << title << '\n';
    for (const Record& record : records) {
        cout << "- score=" << record.score << ", name=" << record.name << '\n';
    }
    cout << '\n';
}

int main() {
    vector<Record> records = {{90, "Ana"}, {75, "Bob"}, {90, "Carla"}, {75, "Diego"}, {90, "Emma"}};

    printRecords(records, "Original order:");

    vector<Record> unstable = records;
    sort(unstable.begin(), unstable.end(),
         [](const Record& a, const Record& b) { return a.score < b.score; });
    printRecords(unstable, "After std::sort (not guaranteed stable):");

    vector<Record> stable = records;
    stable_sort(stable.begin(), stable.end(),
                [](const Record& a, const Record& b) { return a.score < b.score; });
    printRecords(stable, "After std::stable_sort:");

    cout << "Notice how records with equal scores keep relative order with stable_sort.\n";
    return 0;
}

File: languages/go/02-core/sorting-and-searching/example/stable-vs-unstable-sort.go
// This helper example focuses on contrasting two sorting behaviors without extra setup noise.

// This extra example extends sorting-and-searching with tie-order comparisons.

package main

import (
	"fmt"
	"sort"
)

// Keep this helper separate so the main example can focus on the larger idea without extra noise.
type record struct {
	score int
	name  string
}

func printRecords(records []record, title string) {
	fmt.Println(title)
	for _, current := range records {
		fmt.Printf("- score=%d, name=%s\n", current.score, current.name)
	}
	fmt.Println()
}

func main() {
	records := []record{
		{score: 90, name: "Ana"},
		{score: 75, name: "Bob"},
		{score: 90, name: "Carla"},
		{score: 75, name: "Diego"},
		{score: 90, name: "Emma"},
	}

	printRecords(records, "Original order:")

	unstable := append([]record(nil), records...)
	sort.Slice(unstable, func(left int, right int) bool {
		return unstable[left].score < unstable[right].score
	})
	printRecords(unstable, "After sort.Slice (not guaranteed stable):")

	stable := append([]record(nil), records...)
	sort.SliceStable(stable, func(left int, right int) bool {
		return stable[left].score < stable[right].score
	})
	printRecords(stable, "After sort.SliceStable:")

	fmt.Println("Notice how equal scores keep their relative order with SliceStable.")
}
